ouroboros-consensus-test/bench-storage/Bench/Ouroboros/Consensus/Storage/LedgerDB/HD.hs

```diff
@@ -379,29 +450,331 @@ data CmdGenResult k v = CmdGenResult {
   Command generator: Main generators
 -------------------------------------------------------------------------------}
 
+-- | Mode for the top-level generator.
+--
+-- * Bench: Generate commands in deterministic order according to the
+--   @'directives'@ configuration parameter, and do not modify the passed in
+--   configuration parameters.
+-- * Test: Generate commands in non-deterministic order, modify the passed in
+--   configuration to use sized parameters.
+data GenMode = BenchMode | TestMode
+
 -- | A stateful generator for sequences of commands.
 --
 -- Note: We return the @'Model'@ because we want to perform invariant and/or
 -- sanity checks on it.
-generator :: GenConfig -> Gen (CmdGenResult k v)
-generator conf = do
+generator ::
+     (Ord k, Arbitrary k, Arbitrary v)
+  => GenConfig
+  -> GenMode
+  -> Gen (CmdGenResult k v)
+generator preConf gmode = do
+  conf <- mkConf
   m <- genInitialModel conf
   ((), m') <- runCmdGen warmup conf m
-  (cmds, m'') <- runCmdGen genCmds conf m'
+  (cmds, m'') <- runCmdGen gen conf m'
   pure $ CmdGenResult {
       cmds
     , ds0 = diffs m'
     , finalModel = m''
     }
-
--- | Generate a list of commands.
-genCmds :: CmdGen k v [Cmd MockDiffSeq k v]
-genCmds = error "genCmds: not implemented"
+  where
+    gen = case gmode of
+      BenchMode -> genCmdsDet
+      TestMode  -> genCmdsNonDet
+
+    -- Replace configuration parameters by QuickCheck sized ones if the
+    -- generatore mode is set to @Test@.
+    mkConf = case gmode of
+      BenchMode -> pure preConf
+      TestMode  -> do
+        n <- getPositive <$> arbitrary
+        k :: Int <- getSmall . getPositive <$> arbitrary
+        pure $ preConf {
+            nrCommands        = n
+          , securityParameter = k
+          , directives        = Nothing
+          , rollbackConfig    = (rollbackConfig preConf) {
+                distribution = Nothing
+              }
+          }
+
+-- | Generate a list of commands in deterministic order.
+--
+-- Note: The order of commands is deterministic, the inputs for these commands
+-- are often semi-randomly generated.
+genCmdsDet ::
+     forall k v. (Ord k, Arbitrary k, Arbitrary v)
+  => CmdGen k v [Cmd MockDiffSeq k v]
+genCmdsDet = do
+  nrc <- reader nrCommands
+  nrg <- gets nrGenerated
+
+  if nrg >= nrc then
+    pure []
+  else do
+    dirsMay <- reader directives
+
+    case dirsMay of
+      Nothing       -> error "No directives given."
+      Just []       -> error "Directives should be infinitish length"
```

Comment:
I don't get why this is an error case.

Reply:
I hope this comment clarifies why:
```haskell
    -- | An infinite list of directives that determines which commands to
    -- generate.
    --
    -- If we run out of directives without generating @nrCommands@ commands, we
    -- would not know how to proceed generating more commands. Therefore, since
    -- @nrCommands@ can be an arbitrary number, we would want the number of
    -- directives to be /at least/ `nrCommands`.  To ensure that we never run
    -- out of directives, the safest bet is to make the list of directives
    -- infinite, for example by defining it to be an infinitely repeating
    -- series.
    --
    -- @Nothing@ signals that we do not want to use any directives. @Just []@
    -- should result in an error, for the reason we mentioned in the paragraph
    -- above.
  , directives        :: Maybe [Directive]
```

Reply:
I also update the error description to be:
```haskell
      Just []       -> error "Unexpectedly ran out of directives."
```